dace/transformation/interstate/move_loop_into_map.py

```diff
@@ -48,39 +51,57 @@ def can_be_applied(self, graph, expr_index, sdfg, permissive=False):
         if body != body_end:
             return False
 
+        # Body must have only a single connected component
+        # NOTE: This is a strict check that can be potentially relaxed.
+        # If only one connected component includes a Map and the others do not create RW dependencies, then we could
+        # proceed with the transformation. However, that would be a case of an SDFG with redundant computation/copying,
+        # which is unlikely after simplification transformations. Alternatively, we could try to apply the
+        # transformation to each component separately, but this would require a lot more checks.
+        if len(list(nx.weakly_connected_components(body._nx))) > 1:
+            return False
+
         # Check if body contains exactly one map
         maps = [node for node in body.nodes() if isinstance(node, nodes.MapEntry)]
         if len(maps) != 1:
             return False
 
-        # Check that everything else is independent of the loop's itervar
-        subgraph = body.scope_subgraph(maps[0])
-        map_exit = body.exit_node(maps[0])
-        descs: Set[str] = set()
+        map_entry = maps[0]
+        map_exit = body.exit_node(map_entry)
+        subgraph = body.scope_subgraph(map_entry)
+        read_set, write_set = body.read_and_write_sets()
+
+        # Check for iteration variable in map and data descriptors
+        if str(itervar) in map_entry.free_symbols:
+            return False
+        for arr in (read_set | write_set):
+            if str(itervar) in set(map(str, sdfg.arrays[arr].free_symbols)):
+                return False
+
+        # Check that everything else outside the Map is independent of the loop's itervar
         for e in body.edges():
-            if not e.data.is_empty():
-                descs.add(e.data.data)
             if e.src in subgraph.nodes() or e.dst in subgraph.nodes():
                 continue
-            if e.dst is maps[0] and isinstance(e.src, nodes.AccessNode):
+            if e.dst is map_entry and isinstance(e.src, nodes.AccessNode):
                 continue
             if e.src is map_exit and isinstance(e.dst, nodes.AccessNode):
                 continue
             if str(itervar) in e.data.free_symbols:
                 return False
+            if isinstance(e.dst, nodes.AccessNode) and e.dst.data in read_set:
+                # NOTE: This is strict check that can be potentially relaxed.
```

Comment:
`if not permissive` can be used here

Reply:
No, this check is too important to be left to the "permissive" flag. Otherwise, I would have used that in the first place.